orm/src/main/java/org/hibernate/search/batchindexing/impl/IdentifierProducer.java

```diff
@@ -93,21 +93,30 @@ public void run(StatelessSession upperSession) throws Exception {
 		log.trace( "finished" );
 	}
 
-	private void inTransactionWrapper(StatelessSession upperSession) throws Exception {
-		StatelessSession session = upperSession;
+	private void inTransactionWrapper(StatelessSessionImpl upperSession) throws Exception {
+		StatelessSessionImpl session = upperSession;
 		if ( upperSession == null ) {
 			if ( tenantId == null ) {
-				session = sessionFactory.openStatelessSession();
+				session = (StatelessSessionImpl) sessionFactory.openStatelessSession();
 			}
 			else {
-				session = sessionFactory.withStatelessOptions().tenantIdentifier( tenantId ).openStatelessSession();
+				session = (StatelessSessionImpl) sessionFactory.withStatelessOptions().tenantIdentifier( tenantId ).openStatelessSession();
 			}
 		}
 		try {
-			Transaction transaction = session.getTransaction();
-			transaction.begin();
-			loadAllIdentifiers( session );
-			transaction.commit();
+			Transaction transaction = session.accessTransaction();
+			final boolean controlTransactions = ! transaction.isActive();
```

Comment:
If you need this pattern several times, you could extract in a separate method, using Lambda style:

    runInTransactionIfNeeded( session, session -> { loadAllIdentifiers( session ); } );